design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,189 @@
+* Any current Jetty specific plugins (e.g. security plugins) will be migrated to Vert.x
+* Vert.x has great support for working with various different network protocols and has has [unrivalled performance/scalability](https://www.techempower.com/benchmarks/)
+characteristics for a JVM toolkit. It will also set us up well for a fully async / reactive internal threading model in the server that we
+should aim towards for the future.
+
+### Implementation plan
+
+There is a lot of work to do here and it should be split up into separate work packages rather than trying to complete it as a single
+monolithic piece of work. Some obvious partitions of the work would be:
+
+* Protocol implementation
+* Java client implementation. This can be developed using a fake server so there is no dependency on completing the server work first.
+* JavaScript client implementation. Again, this can be developed using a fake server
+* Server streaming implementation
+* Migration of existing endpoints to new server implementation
+* Migration of existing plugins to new server implementation.
+* Switchover from old API to new API
+
+We should retain the current API implementation until it's ready to be switched over to the new implementation. The new implementation can exist
+in the same codebase during development, and only after switchover will the old implementation be removed.
+That way we can ensure a seamless switchover as there is a working API at all times.
+
+## Test plan
+
+We will require unit/module level tests and integration tests for all the new or changed components as per standard best practice.
+
+## Documentation Updates
+
+* We will produce new guide(s) for the Java and JavaScript clients.
+* We will produce example applications showing how to use the client in a real app. E.g. using Spring Boot / Vert.x (Java) and Node.js (JavaScript)
+* The current documentation for the HTTP/REST API can remain more or less as-is with the removal of the problematic parts.
+* We will write up a document on the new wire protocol to encourage 3rd parties to implement new clients.
+* There may be some server side configuration changes due to the new server implementation.
+
+# Compatibility Implications
+
+The current chunked response query streaming endpoint will be removed so any users currently using that will have to upgrade.
+
+The current websockets query streaming endpoint will be removed so any users currently using that will have to upgrade.
+This endpoint is currently undocumented so it's not expected a lot of users are using it. It is used by C3 (?)
+so that will need to be migrated to the new API.
+
+## Performance Implications
+
+* Streaming query results with the new client and protocol should provide significantly better performance than
+the current websockets or HTTP streaming APIs
+* Using Vert.x in general for hosting the API will provide excellent performance and scalability with very low resource usage.
```

Comment:
For my understanding, this would come from less sockets used and multiplexing?

Reply:
The advantages here come more from the Vert.x non blocking threading model which means more work can be done with fewer threads so less RAM and context switching. And also from the fact that both Vert.x and Netty are very highly optimised so simply perform better than pretty much any other offering in the JVM world.
Having said that we might not notice it if we're only handling a small amount of traffic.

Reply:
do you have any prior benchmarks for vert.x vs jersey/glassfish- the stack we use? or anything similar.. this is for my own understanding.

Reply:
One good source is Techempower benchmarks https://www.techempower.com/benchmarks/
It's probably the best known set of independent benchmarks in the industry, and pretty much every framework in the industry (not just JVM) is covered. They've been running for quite a few years.